### PR TITLE
adding support for SSO based on auth2-proxy

### PR DESCRIPTION
**Pull request by rkachach**
This PR builds on top of the two previous ones: https://github.com/ceph/ceph/pull/57535 and https://github.com/ceph/ceph/pull/58402. It inroduces the new service `oauth2-proxy` (the open source project) which provides OIDC support out of the box. In combination with the `mgmt-gateway` this service provides a seamless SSO to access all the Ceph managment backend services (Ceph Dashboard and monitoring stack). 

**Changes included in this PR**:
- [x]  Introducing `oauth2-proxy` new service
- [x]  Add support for Grafana seamless SSO login (based on oauth2)
- [x] Change how we get fqdn for hosts (use inventory information)
- [x] Refactoring of Grafana daemon configuration code

Following diagram depicts the new architecture:

Following are the steps to setup working environment in this case using `dex` idp for testing purpose:

**1) Run dex with the following config file:**

```
issuer: http://<ip-of-dex-node>:5556/dex

storage:
  type: sqlite3
  config:
    file: /var/dex/dex.db

web:
  http: 0.0.0.0:5556

expiry:
  deviceRequests: 5h
  signingKeys: 6h
  idTokens: 24h
  authRequests: 24h

logger:
  level: info
  format: text

oauth2:
  responseTypes: [code]
  skipApprovalScreen: false
  alwaysShowLoginScreen: false

enablePasswordDB: true

staticClients:
  - id: oauth2-proxy
    redirectURIs:
      - 'https://<ip-of-mgmt-gw-node>/oauth2/callback'
    name: 'oauth2-proxy'
    secret: proxy

connectors:
  - type: mockCallback
    id: mock
    name: Example

staticPasswords:
  - email: "admin@example.com"
    hash: "$2a$10$2b2cU8CPhOTaGrs1HRQuAueS7JTT5ZHsHSzYiFPm1leZck7Mc8T4W"
    username: "admin"
    userID: "08a8684b-db88-4b73-90a9-3cd1661f5466"
```
**run dex server:**

```
podman run --rm -p 5556:5556 -p 5555:5555 -v /<path-to-your-config.yaml:/etc/dex/config.docker.yaml --name dex-container ghcr.io/dexidp/dex
```

**2) Configure mgmt-gateway service**

```
ceph orch apply mgmt-gateway --enable_auth=true
```

Wait for all the services to get reconfigured (it takes less than 1min)

**3) Configure oauth2-proxy service**

```
service_type: oauth2-proxy
placement:
  hosts:
    - ceph-node-1
spec:
  https_address: "0.0.0.0:4180"
  provider_display_name: "My OIDC Provider"
  client_id: "oauth2-proxy"
  client_secret: "proxy"
  oidc_issuer_url: "http://<your-ip>:5556/dex"
```
Deploy the service:

```
ceph orch apply -i auth2-proxy.yaml
```

Wait for all the services to get reconfigured (it takes less than 1min)

**4) Open a browser and visit `https://<mgmt-gw-node-ip>`**

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [x] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: doc/cephadm/services/oauth2-proxy.rst [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+==================
+
+Deploying oauth2-proxy
+======================
+
+In Ceph releases starting from Squid, the `oauth2-proxy` service introduces an advanced method
+for managing authentication and access control for Ceph applications. This service integrates
+with external Identity Providers (IDPs) to provide secure, flexible authentication via the
+OIDC (OpenID Connect) protocol. `oauth2-proxy` acts as an authentication gateway, ensuring that
+access to Ceph applications such as the Ceph Dashboard and monitoring stack tools is tightly controlled.
```
Comment (anthonyeleven) [inline doc/cephadm/services/oauth2-proxy.rst:14]: s/such as/including/
s/tools//

Comment (rkachach) [reply doc/cephadm/services/oauth2-proxy.rst]: fixed

**Comment by rkachach** [comment]
jenkins test make check

**Comment by zdover23** [comment]
The docs portion of this looks mergeable to me, but I don't feel comfortable approving it since it contains much more than mere documentation updates.

**Comment by rkachach** [comment]
> The docs portion of this looks mergeable to me, but I don't feel comfortable approving it since it contains much more than mere documentation updates.

No worries, that's normal. If you are OK with docs you can give your approval indicating that it's for the docs part. Anyway this PR cannot be merged without an approval from orchesrtation team.

**Comment by rkachach** [comment]
jenkins test rook e2e

**Comment by rkachach** [comment]
jenkins test dashboard cephadm

**Comment by rkachach** [comment]
jenkins test rook e2e

**Comment by rkachach** [comment]
jenkins test dashboard cephadm

**Comment by rkachach** [comment]
jenkins test make check

**Comment by rkachach** [comment]
jenkins test dashboard

**Comment by rkachach** [comment]
jenkins test make check

**Comment by rkachach** [comment]
jenkins test rook e2e

**Comment by rkachach** [comment]
jenkins test dashboard cephadm

**Comment by rkachach** [comment]
jenkins test api

**Comment by rkachach** [comment]
jenkins test dashboard cephadm

**Comment by rkachach** [comment]
jenkins test dashboard

**Comment by rkachach** [comment]
jenkins test api

**Comment by rkachach** [comment]
jenkins test dashboard cephadm

**Comment by rkachach** [comment]
jenkins test api

**Comment by rkachach** [comment]
jenkins test dashboard

**Comment by rkachach** [comment]
jenkins test dashboard cephadm

**Comment by rkachach** [comment]
jenkins test dashboard

**Comment by rkachach** [comment]
jenkins test api

**Comment by rkachach** [comment]
jenkins test rook e2e

**Comment by rkachach** [comment]
jenkins test rook e2e

**Comment by rkachach** [comment]
jenkins test rook e2e

File: src/python-common/ceph/deployment/service_spec.py [hunk 2]
```diff
@@ -1906,6 +1911,118 @@ def _validate_ssl_protocols(self, protocols: Optional[List[str]]) -> None:
...
+        hostname, port = result.netloc.rsplit(':', 1)
+        # Validate port
+        if not port.isdigit() or not (0 <= int(port) <= 65535):
+            raise SpecValidationError("Invalid https_address: Port must be between 0 and 65535.")
+
+    def _validate_cookie_secret(self, cookie_secret: Optional[str]) -> None:
+        if cookie_secret is None:
+            return
+        if not isinstance(cookie_secret, str):
+            raise SpecValidationError("Invalid cookie_secret: Must be a non-empty string.")
```
Comment (adk3798) [inline src/python-common/ceph/deployment/service_spec.py:2006]: does this check actually catch the empty string case?

**Review by adk3798 — APPROVED** [review]
LGTM, not really any complaints about the code, although I'm not really able to verify if the configs being built are correct. Is there some way we could come up with an integration test for this since it's such a large feature?

**Comment by rkachach** [comment]
jenkins test rook e2e

**Comment by adk3798** [comment]
https://pulpito.ceph.com/adking-2024-08-10_01:02:35-orch:cephadm-wip-adk-testing-2024-08-09-1319-distro-default-smithi/

reruns: https://pulpito.ceph.com/adking-2024-08-11_18:59:43-orch:cephadm-wip-adk-testing-2024-08-09-1319-distro-default-smithi/

Failures: 
- tasks/nfs failure was actually caused by a PR in the run
- 4x mds_upgrade_sequence failure, known issue
- staggered upgrade ceph versions vs. ceph orch ps mismatch, known issue
- test_monitoring_stack test, broke when updating monitoring stack container images recently, known issue

Run seemed all good other than for the one PR causing the nfs failures

**Comment by rkachach** [comment]
jenkins retest this please

File: doc/cephadm/services/oauth2-proxy.rst [hunk 3]
```diff
@@ -0,0 +1,136 @@
+.. _deploy-cephadm-oauth2-proxy:
+
+==================
+OAuth2 Proxy
+==================
+
+Deploying oauth2-proxy
+======================
+
+In Ceph releases starting from Squid, the `oauth2-proxy` service introduces an advanced method
```
[the remark this answers is not in the document]
Comment (rkachach) [reply doc/cephadm/services/oauth2-proxy.rst:10]: @adk3798 I agree, what alternative do you propose in this case? something like `19.x`?

File: doc/cephadm/services/oauth2-proxy.rst [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    placement:
+      hosts:
+        - ceph0
+    spec:
+     https_address: "0.0.0.0:4180"
+     provider_display_name: "My OIDC Provider"
+     client_id: "your-client-id"
+     oidc_issuer_url: "http://192.168.100.1:5556/dex"
+     client_secret: "your-client-secret"
+     cookie_secret: "your-cookie-secret"
```
[the remark this answers is not in the document]
Comment (rkachach) [reply doc/cephadm/services/oauth2-proxy.rst:81]: I added a link to the `oauth2-proxy` project official web page

File: src/cephadm/cephadmlib/daemons/oauth2_proxy.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+                        'required file missing from config-json: %s' % fname
+                    )
+
+    @staticmethod
+    def get_version(
+        ctx: CephadmContext, fsid: str, daemon_id: str
+    ) -> Optional[str]:
+        """Return the version of the notifier from it's http endpoint"""
+        # Redo(TODO): fix version
+        return 'TODO'
```
[the remark this answers is not in the document]
Comment (rkachach) [reply src/cephadm/cephadmlib/daemons/oauth2_proxy.py]: Fixed

File: src/cephadm/cephadmlib/daemons/oauth2_proxy.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+
+    def customize_container_args(
+        self, ctx: CephadmContext, args: List[str]
+    ) -> None:
+        uid, _ = self.uid_gid(ctx)
+        other_args = [
+            '--user',
+            str(uid),
+            # FIXME: disable cpu/memory limits for the time being (not supported
+            # by ubuntu 18.04 kernel!)
```
[the remark this answers is not in the document]
Comment (rkachach) [reply src/cephadm/cephadmlib/daemons/oauth2_proxy.py:160]: removed, just a left over comment

File: src/python-common/ceph/deployment/service_spec.py (hunk 2 above)
Comment (rkachach) [reply src/python-common/ceph/deployment/service_spec.py:2007]: this code was changed in a later commit, it should detect all the cases

**Comment by rkachach** [comment]
jenkins test rook e2e

**Comment by rkachach** [comment]
jenkins test api

**Comment by rkachach** [comment]
jenkins retest this please

**Comment by rkachach** [comment]
jenkins retest this please

**Comment by rkachach** [comment]
jenkins test rook e2e

**Comment by rkachach** [comment]
jenkins test dashboard

**Comment by rkachach** [comment]
jenkins test dashboard

**Comment by rkachach** [comment]
jenkins test rook e2e

**Comment by rkachach** [comment]
jenkins test rook e2e

**Comment by rkachach** [comment]
jenkins test dashboard cephadm

**Comment by rkachach** [comment]
jenkins test dashboard

**Comment by rkachach** [comment]
jenkins test dashboard

**Comment by rkachach** [comment]
jenkins test rook e2e

**Comment by afreen23** [comment]
Hello,

I tested this using keycloak as IdP, I tried adding "redirect_url" in config itself and it came up as expected.

Thanks

**Comment by rkachach** [comment]
> Hello,
> 
> I tested this using keycloak as IdP, I tried adding "redirect_url" in config itself and it came up as expected.
> 
> Thanks

Thanks Afreen, latest code should add that field automatically :+1:

**Comment by rkachach** [comment]
jenkins test rook e2e

**Comment by rkachach** [comment]
jenkins test rook e2e

**Comment by rkachach** [comment]
jenkins test dashboard

**Comment by rkachach** [comment]
jenkins test dashboard cephadm

**Comment by rkachach** [comment]
jenkins test dashboard

**Pull request merged**